great_expectations/data_context/data_context.py

```diff
@@ -1723,82 +1744,119 @@ def profile_datasource(self,
+                profiling_results = {
+                    'success': False,
+                    'error': {
+                        'code': DataContext.PROFILING_ERROR_CODE_SPECIFIED_DATA_ASSETS_NOT_FOUND,
+                        'not_found_data_assets': not_found_data_assets,
+                        'data_assets': data_asset_name_list
+                    }
+                }
+                return profiling_results
+
+
+            data_asset_name_list = data_assets
             data_asset_name_list.sort()
-            data_asset_name_list = data_asset_name_list[:max_data_assets]
+            total_data_assets = len(data_asset_name_list)
+            if not dry_run:
+                logger.info("Profiling the white-listed data assets: %s, alphabetically." % (",".join(data_asset_name_list)))
+        else:
+            if profile_all_data_assets:
+                data_asset_name_list.sort()
+            else:
+                if total_data_assets > max_data_assets:
+                    profiling_results = {
+                        'success': False,
+                        'error': {
+                            'code': DataContext.PROFILING_ERROR_CODE_TOO_MANY_DATA_ASSETS,
+                            'num_data_assets': total_data_assets,
+                            'data_assets': data_asset_name_list
+                        }
+                    }
+                    return profiling_results
 
-        total_columns, total_expectations, total_rows, skipped_data_assets = 0, 0, 0, 0
-        total_start_time = datetime.datetime.now()
-        run_id = total_start_time.isoformat().replace(":", "") + "Z"
-        for name in data_asset_name_list:
-            logger.info("\tProfiling '%s'..." % name)
-            try:
-                start_time = datetime.datetime.now()
+        if not dry_run:
+            logger.info("Profiling all %d data assets from generator %s" % (len(data_asset_name_list), generator_name))
+        else:
+            logger.info("Found %d data assets from generator %s" % (len(data_asset_name_list), generator_name))
 
-                # FIXME: There needs to be an affordance here to limit to 100 rows, or downsample, etc.
-                batch = self.get_batch(
-                    data_asset_name=NormalizedDataAssetName(datasource_name, generator_name, name),
-                    expectation_suite_name=profiler.__name__
-                )
+        profiling_results['success'] = True
 
-                if not profiler.validate(batch):
-                    raise ProfilerError(
-                        "batch '%s' is not a valid batch for the '%s' profiler" % (name, profiler.__name__)
+        if not dry_run:
+            profiling_results['results'] = []
+            total_columns, total_expectations, total_rows, skipped_data_assets = 0, 0, 0, 0
+            total_start_time = datetime.datetime.now()
+            run_id = total_start_time.isoformat().replace(":", "") + "Z"
```

Comment:
Since profiling is now contemplated as a multi-step option (with different data assets specified in different runs), this run_id should be updated to be something reflecting the profiling behavior